onboarding/mlsys/1-backprop.md

```diff
@@ -0,0 +1,414 @@
+take `m` and a single `(x, ylabel)` pair as inputs and return.
+
+Then write a function to calculate the total gradient for a list of
+`(x, ylabel)` pairs.
+
+Is it possible to write the second function by making multiple calls to the
+first function and combining the results?  Why or why not?
+
+## Autodifferentiation
+
+We solved our gradient symbolically because it was so simple.  But real
+gradients in real models are calculated with some magic called
+"autodifferentiation".
+
+You don't need to understand autodifferentiation, but you do need to
+understand:
+- it starts in the the "forward" pass of training a batch, and finishes in
+  the "backward" pass, and
+- it is additive.
+
+The "forward" pass is when a model makes a prediction, and the loss is
+calculated by comparing the label to the prediction.  During the forward pass,
+the deep learning library keeps information on how every weight in the model
+affects the output.  Then the "backward" pass takes the calculated loss and
+figures out how chaning each weight would affect the loss.  This output is the
+gradient.  In our toy model, the gradient is just a single number because we
+have just a single parameter (`m`), but in a model with a billion parameters the
+gradient is a billion numbers long.
+
+The additive part is important.  If you do a forward pass on a particular data
+point, follwed by a backward pass, you'll have some gradient.  If you
+immediately do another forward pass on the same data, followed by another
+background pass, you'll have twice the gradient as before.  If you do it one
+more time, you'll have three times the original gradient.
+
+### Q4:
+
+The additive nature of autodifferentiated gradients is important to
+understanding the features that Determined offers.
+
+Explain in your own words _why_ autodifferentiated gradients are additive.
+
+## Training by Gradient Descent
+
+The basic gradient desent loop algorithm is:
+
+1. Pick an initial model weight.
+2. Select one or more `(x, ylabel)` pairs from the dataset.
+3. Calculate a prediction for the selected inputs with a label (Q2).
+4. Calculate the gradient using the prediction vs label (Q3).
+5. Subtract `gradient * lr` from our model weight ("Minimizing the loss").
+6. Repeat steps 2—5 until you decide to stop.
+
+### Q5
+
+Write a training loop to do gradient descent on our `ypred = mx` model.  Your
+function should take the following inputs:
+
+- initial model weight `m`
+- a learning rate `lr` for calculating the setp
+- a dataset, which is a list of `(x, ylabel)` data pairs
+- a number of iterations to do (one "iteration" is steps 2-5).
+
+Your function should return the final trained weight `m`.
+
+Additionally, in the "select one or more `(x, ylabel)` input pairs" step, for
+now you should just select the next `(x, ylabel)` pair in the dataset, and it
+should wrap back to the previous dataset.  So if your dataset is of length 10,
```

Comment:
> should wrap back to the previous dataset

Do you want to introduce shuffling here? Doesn't matter for the example at hand.

Reply:
I don't, but I'll put that on my todo list.

Reply:
I decided in my rework that shuffling was more of a higher-order 'how do you get good training results' thing and I didn't want it in this doc.